### PR TITLE
Capsule/Capsule distance function: reimplementation

### PR DESCRIPTION
**Pull request by rstrudel**

### CODE REVIEW

File: src/distance_capsule_capsule.cpp [hunk 1]
```diff
@@ -25,299 +27,121 @@ namespace hpp
...
+    FCL_REAL halfLength1 = capsule1->halfLength;
+    FCL_REAL halfLength2 = capsule2->halfLength;
+    FCL_REAL radius1 = capsule1->radius;
+    FCL_REAL radius2 = capsule2->radius;
+    // direction of capsules
+    // ||d1|| = 2 * halfLength1
+    // Matrix3f::ConstColXpr d1 = 2 * halfLength1 * tf1.getRotation().col(2);
+    // Matrix3f::ConstColXpr d2 = 2 * halfLength2 * tf2.getRotation().col(2);
+    const fcl::Vec3f& d1 = 2 * halfLength1 * tf1.getRotation().col(2);
+    const fcl::Vec3f& d2 = 2 * halfLength2 * tf2.getRotation().col(2);
```
Comment (jmirabel) [inline src/distance_capsule_capsule.cpp]: The ampersands should be removed.

**Comment by rstrudel** [comment]
I encountered some bugs while trying the Capsule/Capsule collision detection. Additional tests have been written by Justin to check the collision function more thoroughly. The previous function did not pass these tests, thus I reimplemented the function using the book [Real-Time Collision Detection, Christer Ericson, p148-149](http://www.r-5.org/files/books/computers/algo-list/realtime-3d/Christer_Ericson-Real-Time_Collision_Detection-EN.pdf).

File: src/distance_capsule_capsule.cpp [hunk 2]
```diff
@@ -25,299 +27,121 @@ namespace hpp
...
+    else if (e <= EPSILON)
+    {
+      // Second segment is degenerated
+      s = CLAMP(-c / a, 0.0, 1.0);
+      t = 0.0;
+    }
+    else
+    {
+      // Always non-negative, equal 0 if the segments are parallel
+      FCL_REAL denom = a*e-b*b;
```
Comment (jmirabel) [inline src/distance_capsule_capsule.cpp]: Maybe the following would be more robust to numerical issues.
```
FCL_REAL denom = std::max(0, a*e-b*b);
```

File: src/distance_capsule_capsule.cpp [hunk 3]
```diff
@@ -25,299 +27,121 @@ namespace hpp
...
+
+      if (denom != 0.0)
+      {
+        s = CLAMP((b*f-c*e) / denom, 0.0, 1.0);
```
Comment (jmirabel) [inline src/distance_capsule_capsule.cpp]: CLAMP suffices here to make the division by denom robust to small values of denom, isn't it ?

**Comment by jcarpent** [comment]
@jmirabel I will make a new release after this PR has been merged.

**Comment by jcarpent** [comment]
I will put some comments here and @rstrudel is free to add his feedback.
@rstrudel realized that the current implementation of distance<Capsule, Capsule> was leading to incorrect behaviors when checking trivial situations as the ones reported in the unit tests. Currently, the previous implementation was not sufficiently tested in the tests. Maybe much more through the hpp pipeline.

Then, @rstrudel decided to implement the segment-to-segment distance check. Which seems appropriate for the Capsule-Capsule test, and in addition, leading to fewer codes. In addition, the reimplementation made by @rstrudel allows removing some constraints on the code (copy past from external lib).

**Comment by jcarpent** [comment]
@rstrudel is currently using the collision between inside his RL framework. So, in a few days, we will see if there is still some bugs, hopefully not.

**Comment by jmirabel** [comment]
> the previous implementation was not sufficiently tested in the tests.

definitely.

> leading to fewer codes.

I saw this and the code is a lot easier to understand.

**Comment by jmirabel** [comment]
I double checked and I think capsule in HPP  are mostly used by @dbury.

**Comment by jmirabel** [comment]
@DianeBury, do you have any comments on this topic ?

**Comment by rstrudel** [comment]
Thanks for accepting the PR. @jcarpent summed it well, I am using hpp-fcl collision checker in my RL framework and I could observe bugs with capsules which lead to this PR. I will use it on a frequently in the following weeks and will see if it has the expected behavior.

**Pull request merged**